File: Modules/ServerTypes/EmployeeProfile/EmployeeProfileForm.ts
import { DateEditor, ImageUploadEditor, StringEditor, EnumEditor, IntegerEditor, LookupEditor, DecimalEditor, MultipleImageUploadEditor, BooleanEditor, PrefixedContext } from "@serenity-is/corelib";
import { SexType } from "./SexType";
import { EmployeeType } from "./EmployeeType";
import { MaritalStatus } from "./MaritalStatus";
import { ProbationClass } from "./ProbationClass";
import { EmployeeCareerPathEditor } from "@/HumanResource/EmployeeProfile/EmployeeCareerPath/EmployeeCareerPathEditor";
import { EmployeeCp38Editor } from "@/HumanResource/EmployeeProfile/EmployeeCp38/EmployeeCp38Editor";
import { EmployeeAllowanceEditor } from "@/HumanResource/EmployeeProfile/EmployeeAllowance/EmployeeAllowanceEditor";
import { FixedDeductionEditor } from "@/HumanResource/EmployeeProfile/FixedDeduction/FixedDeductionEditor";
import { EPFClass } from "./EPFClass";
import { EISClass } from "./EISClass";
import { SOCSOClass } from "./SOCSOClass";
import { HRDFClass } from "./HRDFClass";
import { initFormType } from "@serenity-is/corelib/q";

export interface EmployeeProfileForm {
    ResignationDate: DateEditor;
    ResignLeaveDate: DateEditor;
    TerminateDate: DateEditor;
    TerminateLeaveDate: DateEditor;
    EmployeeImg: ImageUploadEditor;
    EmployeeName: StringEditor;
    EmployeeID: StringEditor;
    Sex: EnumEditor;
    Birthday: DateEditor;
    EmployeeEmail: StringEditor;
    EmployeeType: EnumEditor;
    RecruitmentDate: DateEditor;
    TelNumber1: StringEditor;
    TelNumber2: StringEditor;
    MaritalStatus: EnumEditor;
    Age: IntegerEditor;
    RetireDate: DateEditor;
    Nric: StringEditor;
    OldNRIC: StringEditor;
    RaceID: LookupEditor;
    NationalityID: LookupEditor;
    JoinDate: DateEditor;
    ProbationPeriod: DecimalEditor;
    ProbationPeriodFrom: StringEditor;
    ProbationPeriodUntil: StringEditor;
    PassedProbation: EnumEditor;
    WorkingPermit: StringEditor;
    ArrivalDate: DateEditor;
    WorkingPermitIssueDate: DateEditor;
    WorkingPermitExpireDate: DateEditor;
    WorkingPermitValidFrom: DateEditor;
    WorkingPermitValidUntil: DateEditor;
    PassportNumber: StringEditor;
    PassportIssueDate: DateEditor;
    PassportValidPeriod: IntegerEditor;
    PassportExpiryDate: DateEditor;
    Address: StringEditor;
    PostcodeId: LookupEditor;
    CountryID: LookupEditor;
    StateID: LookupEditor;
    CityID: LookupEditor;
    DivisionID: LookupEditor;
    DepartmentID: LookupEditor;
    SectionID: LookupEditor;
    JobGradeID: LookupEditor;
    Cp8dID: LookupEditor;
    OccupationID: LookupEditor;
    CostCentreID: LookupEditor;
    JobDescription: MultipleImageUploadEditor;
    EmployeeCareerPath: EmployeeCareerPathEditor;
    CreateUser: BooleanEditor;
    GrantHRPrivilege: BooleanEditor;
    UserName: StringEditor;
    UserPassword: StringEditor;
    UserRowID: IntegerEditor;
    WorkingSpouse: BooleanEditor;
    ChildrenInUniversity: IntegerEditor;
    ChildrenUnderEighteen: IntegerEditor;
    DisabledChildInUniversity: IntegerEditor;
    DisabledChild: IntegerEditor;
    BasicSalary: DecimalEditor;
    DailyWorkingMinute: IntegerEditor;
    NightShiftAllowancePerDay: DecimalEditor;
    BankID: LookupEditor;
    BankAccountNumber: StringEditor;
    Cp38Lists: EmployeeCp38Editor;
    OtPayEntitlement: BooleanEditor;
    FixedOtRateOption: BooleanEditor;
    PayByMonth: BooleanEditor;
    PayByHour: BooleanEditor;
    PayByDay: BooleanEditor;
    AllowanceLists: EmployeeAllowanceEditor;
    FixedDeductionList: FixedDeductionEditor;
    EpfClass: EnumEditor;
    EpfAccountNumber: StringEditor;
    EisClass: EnumEditor;
    SocsoClass: EnumEditor;
    SocsoAccountNumber: StringEditor;
    SsfwEffectiveDate: DateEditor;
    SsfwNumber: StringEditor;
    HRDFClass: EnumEditor;
    PCBnumber: StringEditor;
    CalculationDate: DateEditor;
    WorkingHour: DecimalEditor;
    WorkingDays: DecimalEditor;
    DailyRateBase: DecimalEditor;
    NplRateBase: DecimalEditor;
    OtRateWeekday: DecimalEditor;
    OtRateWeekend: DecimalEditor;
    OtRatePublicHoliday: DecimalEditor;
}

export class EmployeeProfileForm extends PrefixedContext {
    static formKey = 'EmployeeProfile.EmployeeProfile';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!EmployeeProfileForm.init)  {
            EmployeeProfileForm.init = true;

            var w0 = DateEditor;
            var w1 = ImageUploadEditor;
            var w2 = StringEditor;
            var w3 = EnumEditor;
            var w4 = IntegerEditor;
            var w5 = LookupEditor;
            var w6 = DecimalEditor;
            var w7 = MultipleImageUploadEditor;
            var w8 = EmployeeCareerPathEditor;
            var w9 = BooleanEditor;
            var w10 = EmployeeCp38Editor;
            var w11 = EmployeeAllowanceEditor;
            var w12 = FixedDeductionEditor;

            initFormType(EmployeeProfileForm, [
                'ResignationDate', w0,
                'ResignLeaveDate', w0,
                'TerminateDate', w0,
                'TerminateLeaveDate', w0,
                'EmployeeImg', w1,
                'EmployeeName', w2,
                'EmployeeID', w2,
                'Sex', w3,
                'Birthday', w0,
                'EmployeeEmail', w2,
                'EmployeeType', w3,
                'RecruitmentDate', w0,
                'TelNumber1', w2,
                'TelNumber2', w2,
                'MaritalStatus', w3,
                'Age', w4,
                'RetireDate', w0,
                'Nric', w2,
                'OldNRIC', w2,
                'RaceID', w5,
                'NationalityID', w5,
                'JoinDate', w0,
                'ProbationPeriod', w6,
                'ProbationPeriodFrom', w2,
                'ProbationPeriodUntil', w2,
                'PassedProbation', w3,
                'WorkingPermit', w2,
                'ArrivalDate', w0,
                'WorkingPermitIssueDate', w0,
                'WorkingPermitExpireDate', w0,
                'WorkingPermitValidFrom', w0,
                'WorkingPermitValidUntil', w0,
                'PassportNumber', w2,
                'PassportIssueDate', w0,
                'PassportValidPeriod', w4,
                'PassportExpiryDate', w0,
                'Address', w2,
                'PostcodeId', w5,
                'CountryID', w5,
                'StateID', w5,
                'CityID', w5,
                'DivisionID', w5,
                'DepartmentID', w5,
                'SectionID', w5,
                'JobGradeID', w5,
                'Cp8dID', w5,
                'OccupationID', w5,
                'CostCentreID', w5,
                'JobDescription', w7,
                'EmployeeCareerPath', w8,
                'CreateUser', w9,
                'GrantHRPrivilege', w9,
                'UserName', w2,
                'UserPassword', w2,
                'UserRowID', w4,
                'WorkingSpouse', w9,
                'ChildrenInUniversity', w4,
                'ChildrenUnderEighteen', w4,
                'DisabledChildInUniversity', w4,
                'DisabledChild', w4,
                'BasicSalary', w6,
                'DailyWorkingMinute', w4,
                'NightShiftAllowancePerDay', w6,
                'BankID', w5,
                'BankAccountNumber', w2,
                'Cp38Lists', w10,
                'OtPayEntitlement', w9,
                'FixedOtRateOption', w9,
                'PayByMonth', w9,
                'PayByHour', w9,
                'PayByDay', w9,
                'AllowanceLists', w11,
                'FixedDeductionList', w12,
                'EpfClass', w3,
                'EpfAccountNumber', w2,
                'EisClass', w3,
                'SocsoClass', w3,
                'SocsoAccountNumber', w2,
                'SsfwEffectiveDate', w0,
                'SsfwNumber', w2,
                'HRDFClass', w3,
                'PCBnumber', w2,
                'CalculationDate', w0,
                'WorkingHour', w6,
                'WorkingDays', w6,
                'DailyRateBase', w6,
                'NplRateBase', w6,
                'OtRateWeekday', w6,
                'OtRateWeekend', w6,
                'OtRatePublicHoliday', w6
            ]);
        }
    }
}

[SexType, EmployeeType, MaritalStatus, ProbationClass, EPFClass, EISClass, SOCSOClass, HRDFClass]; // referenced types

File: Modules/HumanResource/EmployeeProfile/EmployeeProfile/EmployeeProfileForm.cs
using Serenity.ComponentModel;
using System;
using System.Collections.Generic;
using System.ComponentModel;

namespace HRMSoftware.EmployeeProfile.Forms;

[FormScript("EmployeeProfile.EmployeeProfile")]
[BasedOnRow(typeof(EmployeeProfileRow), CheckNames = true)]
public class EmployeeProfileForm
{
    [Tab("Basic Employee Information")]
    [HideOnUpdate, HideOnInsert]
    public DateTime ResignationDate { get; set; }
    [HideOnUpdate, HideOnInsert]
    public DateTime ResignLeaveDate { get; set; }

    [HideOnUpdate, HideOnInsert]
    public DateTime TerminateDate { get; set; }
    [HideOnUpdate, HideOnInsert]
    public DateTime TerminateLeaveDate { get; set; }


    
    [Category("Basic Info")]
    public string EmployeeImg { get; set; }
  
    [HalfWidth]
    public string EmployeeName { get; set; }
  
    [HalfWidth]
    public string EmployeeID { get; set; }
    
    [HalfWidth]
    public int Sex { get; set; }
    

    //[HideOnInsert]
    //public DateTime LastPaymentDate { get; set; }
    [HalfWidth]
    public DateTime Birthday { get; set; }
    [HalfWidth]
    public string EmployeeEmail { get; set; }

    [HalfWidth]
    public int EmployeeType { get; set; }


    [HalfWidth]
    public DateTime RecruitmentDate { get; set; }

    [HalfWidth]
    public string TelNumber1 { get; set; }


    [HalfWidth]
    public string TelNumber2 { get; set; }




    [HalfWidth]
    public int MaritalStatus { get; set; }


    [HalfWidth]
    public int Age { get; set; }


    [HalfWidth]
    public DateTime RetireDate { get; set; }

    [HalfWidth]
    public string Nric { get; set; }

    [HalfWidth]
    public string OldNRIC    { get; set; }





    [HalfWidth]
    public int RaceID { get; set; }
 
    [HalfWidth]
    public int NationalityID { get; set; }
    public DateTime JoinDate { get; set; }


    [Category("Probation")]


    [OneThirdWidth]
    public double ProbationPeriod { get; set; }

    [OneThirdWidth]
    public string ProbationPeriodFrom { get; set; }

    [OneThirdWidth]
    public string ProbationPeriodUntil { get; set; }

    public int PassedProbation { get; set; }

    [Category("Passport Details")]
    [HalfWidth]
    public string WorkingPermit { get; set; }

    [HalfWidth]
    public DateTime ArrivalDate { get; set; }


    [HalfWidth]
    public DateTime WorkingPermitIssueDate { get; set; }
    [HalfWidth]
    public DateTime WorkingPermitExpireDate { get; set; }
    [HalfWidth]
    public DateTime WorkingPermitValidFrom { get; set; }
    [HalfWidth]
    public DateTime WorkingPermitValidUntil { get; set; }


    
    [HalfWidth]
    public string PassportNumber { get; set; }
    [HalfWidth]
    public DateTime PassportIssueDate { get; set; }
    [HalfWidth]
    public int PassportValidPeriod { get; set; }
    [HalfWidth]
    public DateTime PassportExpiryDate { get; set; }



    [Category("Address")]
    public string Address { get; set; }
    [HalfWidth]
    public int PostcodeId { get; set; }
    [HalfWidth]
    public int CountryID { get; set; }

    [HalfWidth]
    public int StateID { get; set; }

    [HalfWidth]
    public int CityID { get; set; }


    [Tab("Working Information")]
    [Category("Organisation Position")]
    [HalfWidth]
    public int DivisionID { get; set; }

    [HalfWidth]
    public int DepartmentID { get; set; }
    [HalfWidth]
    public int SectionID { get; set; }
    [Category("Position Details")]

    [HalfWidth]
    public int JobGradeID { get; set; }

    [HalfWidth]
    public int Cp8dID { get; set; }
    [HalfWidth]
    public int OccupationID { get; set; }
    [HalfWidth]
    public int CostCentreID { get; set; }

    public string JobDescription { get; set; }
    [Category("Employee Career Path")]
    [IgnoreName, LabelWidth("0"), EmployeeCareerPathEditor]
    public List<EmployeeCareerPathRow> EmployeeCareerPath { get; set; }


    [Category("User Creation")]

    [OneThirdWidth]
    public int CreateUser { get; set; }

    [OneThirdWidth]
    public int GrantHRPrivilege { get; set; }

    [HalfWidth]
    public string UserName { get; set; }

    [HalfWidth]
    public string UserPassword { get; set; }

    [HideOnInsert,HideOnUpdate]
    public int UserRowID { get; set; }


    [Tab("Payments Information")]
    [Category("")]

    [OneThirdWidth]
    public int WorkingSpouse { get; set; }
    [OneThirdWidth]
    public int ChildrenInUniversity { get; set; }
    [OneThirdWidth]
    public int ChildrenUnderEighteen { get; set; }

    [HalfWidth]
    public int DisabledChildInUniversity { get; set; }
    [HalfWidth]
    public int DisabledChild { get; set; }


    [OneThirdWidth]
    public double BasicSalary { get; set; }

    [OneThirdWidth]
    public int DailyWorkingMinute { get; set; }

    [OneThirdWidth]
    public double NightShiftAllowancePerDay { get; set; }


    [HalfWidth]
    public int BankID { get; set; }

    [HalfWidth]
    public string BankAccountNumber { get; set; }
    /*
    [HalfWidth]
    public int SocsoClass { get; set; }

    [HalfWidth]
    public string SocsoAccountNumber { get; set; }



    [HalfWidth]
    public string PcbRate { get; set; }



    [HalfWidth]
    public int EpfClass { get; set; }
    [HalfWidth]
    public string EpfAccountNumber { get; set; }


    [HalfWidth]
    public int EisClass { get; set; }
    //[HalfWidth]
    //public string EISnumber { get; set; }
    */


    [Category("EmployeeCp38")]
    [IgnoreName, LabelWidth("0"), EmployeeCp38Editor]
    public List<EmployeeCp38Row> Cp38Lists { get; set; }
    [Category("Payment Mode")]

    public bool OtPayEntitlement { get; set; }
    public bool FixedOtRateOption { get; set; }
    public int PayByMonth { get; set; }
    public int PayByHour { get; set; }
    public int PayByDay { get; set; }



    [Category("Allowances")]
    [IgnoreName, LabelWidth("0"), EmployeeAllowanceEditor]
    public List<EmployeeAllowanceRow> AllowanceLists { get; set; }

    [Category("Deductions")]
    [IgnoreName, LabelWidth("0"), FixedDeductionEditor]
    public List<FixedDeductionRow> FixedDeductionList { get; set; }



    [Tab("Government Payments")]
    [Category("Epf")]

    [HalfWidth]
    public int EpfClass { get; set; }
    [HalfWidth]
    public string EpfAccountNumber { get; set; }


    [Category("Eis")]

    [MediumHalfLargeQuarterWidth]
    public int EisClass { get; set; }

    [Category("Socso")]


    [HalfWidth]
    public int SocsoClass { get; set; }


    [HalfWidth]
    public string SocsoAccountNumber { get; set; }
    [HalfWidth]
    public DateTime SsfwEffectiveDate { get; set; }


    [HalfWidth]
    public string SsfwNumber { get; set; }


    [Category("HRDF")]


    [HalfWidth]
    public int HRDFClass { get; set; }

    [Category("Pcb")]


    [HalfWidth]
    public string PCBnumber { get; set; }

    [Tab("Rates")]

    [Category("Rates")]
    public DateTime CalculationDate { get; set; }
    public double WorkingHour { get; set; }
    public double WorkingDays { get; set; }
    public double DailyRateBase { get; set; }
    public double NplRateBase { get; set; }
    public double OtRateWeekday { get; set; }
    public double OtRateWeekend { get; set; }
    public double OtRatePublicHoliday { get; set; }

}